ai-prompt-flow/editorextensions/ai-prompt-flow/src/blocks/blocks.ts

```diff
@@ -0,0 +1,653 @@
+                    const {key: name, value} = getKeyValuePairFromRow(row);
+                    return [name, value];
+                }),
+        };
+    }
+
+    public setVariableName(name: string, index: number) {
+        const {keyCell: nameCell} = getKeyValuePairFromTable(index + this.numHeaderRows, this);
+        nameCell.setText(name);
+    }
+
+    public setVariableValue(value: string, index: number) {
+        const {valueCell} = getKeyValuePairFromTable(index + this.numHeaderRows, this);
+        valueCell.setText(value);
+    }
+
+    public addVariable() {
+        addRow(this);
+    }
+
+    public deleteVariable(editableRow: number) {
+        this.deleteRow(editableRow + this.numHeaderRows);
+    }
+}
+
+export class SampleBlock extends TableBlockProxy implements AiBlock {
+    public aiBlockType = AiBlockType.Examples;
+    public static canBeSampleBlock(block: unknown): block is TableBlockProxy {
+        return block instanceof TableBlockProxy && !!block.shapeData.get('isSampleInput');
+    }
+    private readonly numHeaderRows = 2;
+    constructor(id: string, client: EditorClient) {
+        super(id, client);
+        if (!SampleBlock.canBeSampleBlock(this)) {
+            throw new Error('Not a sample block');
+        }
+    }
+
+    private getSystemPrompt() {
+        const systemPrompt = this.getConnectedLines()
+            .map((line) => line.getDownstreamConnection())
+            .filter(isInstanceOf(SystemPromptBlock))[0];
+
+        return systemPrompt?.getPromptText();
+    }
+
+    public toJSON(): SerializedExamplesBlock {
+        const examples: [string, string][] = this.getRows()
+            .slice(this.numHeaderRows)
+            .map((row) => {
+                const {key: name, value} = getKeyValuePairFromRow(row);
+                return [name, value];
+            });
+        return {
+            'id': this.id,
+            'block-type': AiBlockType.Examples,
+            'examples': examples,
+            'system-prompt': this.getSystemPrompt(),
+        };
+    }
+
+    public setUser(text: string, index: number) {
+        const {keyCell: nameCell} = getKeyValuePairFromTable(index + this.numHeaderRows, this);
+        nameCell.setText(text);
+    }
+
+    public setResponse(text: string, index: number) {
+        const {valueCell} = getKeyValuePairFromTable(index + this.numHeaderRows, this);
+        valueCell.setText(text);
+    }
+
+    public addSample() {
+        addRow(this);
+    }
+
+    public deleteSample(editableRow: number) {
+        this.deleteRow(editableRow + this.numHeaderRows);
+    }
+}
+
+function addRow(table: TableBlockProxy, index: number = table.getRows().length - 1) {
+    const {keyCell: cell} = getKeyValuePairFromTable(index, table);
+    const currentStyles = cell.getTextStyle();
+    const newRow = table.addRow(cell, false);
+    for (const newCell of newRow.getCells()) {
+        // See https://lucidsoftware.slack.com/archives/C01BTPAMZNJ/p1695334462696559
```

Comment:
The slack link won't be helpful to anyone outside of Lucid. The tl;dr should be enough by itself.